src/uct/tcp/tcp_ep.c

```diff
@@ -313,22 +296,36 @@ ucs_status_t uct_tcp_ep_create(const uct_ep_params_t *params,
     return status;
 }
 
-void uct_tcp_ep_mod_events(uct_tcp_ep_t *ep, uint32_t add, uint32_t remove)
+void uct_tcp_ep_mod_events(uct_tcp_ep_t *ep, int add, int remove)
 {
-    int old_events = ep->events;
-    int new_events = (ep->events | add) & ~remove;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    int old_events         = ep->events;
+    int new_events         = (ep->events | add) & ~remove;
+    ucs_status_t status;
 
     if (new_events != ep->events) {
         ep->events = new_events;
         ucs_trace("tcp_ep %p: set events to %c%c", ep,
-                  (new_events & EPOLLIN)  ? 'i' : '-',
-                  (new_events & EPOLLOUT) ? 'o' : '-');
+                  (new_events & UCS_EVENT_SET_EVREAD)  ? 'i' : '-',
+                  (new_events & UCS_EVENT_SET_EVWRITE) ? 'o' : '-');
         if (new_events == 0) {
-            uct_tcp_ep_epoll_ctl(ep, EPOLL_CTL_DEL);
+            status = ucs_event_set_del(iface->event_set, ep->fd);
+            if (status != UCS_OK) {
+                ucs_fatal("ucs_event_set_del(fd=%d) failed: %m", ep->fd);
+            }
         } else if (old_events != 0) {
-            uct_tcp_ep_epoll_ctl(ep, EPOLL_CTL_MOD);
+            status = ucs_event_set_mod(iface->event_set, ep->fd,
+                                           ep->events, (void *)ep);
+            if (status != UCS_OK) {
```

Comment:
you can move if (status != UCS_OK) after new events handled, i.e.:
```
if (new_events == 0) {
    status = ucs_event_set_del(iface->event_set, ep->fd);
} else if (old_events != 0) {
    status = ucs_event_set_mod(iface->event_set, ep->fd,
                               ep->events, (void *)ep);
} else {
    status = ucs_event_set_add(iface->event_set, ep->fd,
                               ep->events, (void *)ep);
}
if (status != UCS_OK) {
    ucs_fatal("Unable to modify event set for tcp_ep %p (fd=%d)", ep, ep->fd);
}
```

Reply:
Simplified using your proposal.